File: AdventOfCode-2021/AdventOfCode.Csharp/Solutions/Day01.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AdventOfCode.Csharp.Solutions
{
    // Day 1: Sonar Sweep (https://adventOfCode.com/2021/day/1)
    public class Day01 : IPuzzle
    {
        public string CalculateSolution(Parts part, string inputData)
        {
            var data = inputData.Split(Environment.NewLine).Select(int.Parse).ToList();
            return part switch
            {
                Parts.Part1 => $"{SolvePart1(data)}",
                Parts.Part2 => $"{SolvePart2(data)}",
                _ => throw new ArgumentOutOfRangeException(nameof(part), part, "There are only 2 parts.")
            };
        }

        private static int SolvePart1(IReadOnlyList<int> data)
        {
            var count = 0;
            for (var i = 1; i < data.Count; i++)
            {
                if (data[i] > data[i - 1])
                    count++;
            }
            return count;
        }

        private static int SolvePart2(IReadOnlyList<int> data)
        {
            var count = 0;
            int? prevMeasurementWindow = null;
            for (var i = 2; i < data.Count; i++)
            {
                var windowSum = data[i] + data[i - 1] + data[i - 2];
                if (windowSum > prevMeasurementWindow)
                {
                    count++;
                }
                prevMeasurementWindow = windowSum;
            }
            return count;
        }
    }
}

File: AdventOfCode-2021/AdventOfCode.Fsharp/Solutions/Day01.fs
namespace AdventOfCode.Fsharp.Solutions

open System
open System.Linq
open AdventOfCode.Fsharp.InputData
open System.Collections.Generic

module Day01 = 
    // let testInput = Day01Data.testInput
    let solvePart1 = fun inputData -> 
        inputData
        |> fun (text : string) -> text.Split(Environment.NewLine)
        |> fun lines -> lines.Select(fun line -> line |> int)
        |> fun values -> values.Aggregate((0, Int32.MaxValue), fun (counter, previousValue) value ->
            if value > previousValue then (counter + 1, value)
            else (counter, value))
        |> fun (counter, lastValue) -> counter


    let count3MeasureSlidingWindows (measurements : IEnumerable<int>) : int = 
        let measures = List.ofSeq(measurements)
        let mutable count = 0
        let mutable lastSum = Int32.MaxValue
        for i = 2 to measures.Count()-1 do
            let newSum = measures.Item(i) + measures.Item(i-1) + measures.Item(i-2)
            count <- if lastSum < newSum then count + 1
                     else count
            lastSum <- newSum
        count
    
    let solvePart2 = fun inputData -> 
        inputData
        |> fun (text:string) -> text.Split(Environment.NewLine)
        |> fun lines -> lines.Select(fun line -> line |> int)
        |> fun values -> count3MeasureSlidingWindows values
        |> fun (counter) -> counter
   
